### PR TITLE
GRADLE-1267 add a filteringCharset property to the copy spec

### PR DESCRIPTION
**Pull request by jnizet**
The copy task (and all the other tasks based on a CopySpec) allow filtering the source files to transform them. But the encoding used to read and write the files was always the platform default encoding, which causes problem if, for example, the source files are encoded with UTF8 but the platform default encoding is another one.

This commit allows specifying a file encoding to use when filtering the files using `fileEncoding = 'UTF8'` (for example).

Note that the name `fileEncoding` was chosen for this new property, and not simply `encoding`, because the `Zip` class, which implements `CopySpec` already has an `encoding` property used to encode the file *names* in the archive.

### CODE REVIEW

**Comment by pioterj** [comment]
Thank you for the pull request. It may be a while before we look into it because of limited capacity of Gradle team to deal with pull requests at the moment.

In the meantime, could you sign a [contributor license agreement](https://gradle.org/contributor-license-agreement/)? This is required for legal reasons before any code contributions can be accepted into the Gradle codebase.

**Comment by jnizet** [comment]
I have already contributed several PRs to gradle (see https://github.com/gradle/gradle/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+jnizet+), so the CLA is already signed.

**Comment by pioterj** [comment]
I found it, sorry to bother you. We just didn't have your github nickname in our spreadsheet. Thanks for continuing to contribute.

**Comment by rjernst** [comment]
+1, this is really important as copy filtering (or any feature that needs to inspect bytes) is simply broken right now with cross platform builds between eg windows and *nix.

**Comment by pioterj** [comment]
This looks like a reasonable and needed change and we appreciate your effort. However I'm afraid that adding new properties here and there is not a good way to approach the problem of lacking encoding support in Gradle. We need to take a step back at look at the problem holistically. We should first have a discussion on [Gradle Developer List](https://groups.google.com/forum/#!forum/gradle-dev) and create a [design document](https://github.com/gradle/gradle/tree/master/design-docs) for comprehensive encoding support.

I see that we have multiple contributors interested in this topic. @jnizet @rjernst @AlphaHinex would you be willing to take a stab at such effort?

**Comment by rjernst** [comment]
@pioterj Can you explain more why you think a new setting is not reasonable? It is exactly what is needed, as this should not be something gradle wide (you can already set that by invoking gradle with eg `-Dfile.encoding=UTF-8` to override the system default). This is an issue with the copy task reading a file without specifying the charset (as should always be done, IMO).

**Comment by jnizet** [comment]
I agree with @rjernst. 

BTW, several copy tasks, or even several `from { }` CopySourceSpecs of the same copy task might each need a different encoding. I feel this is exactly what is needed, and it's an important feature that is lacking for a long time (the bug is opened for 5 years).

**Comment by pioterj** [comment]
I'm not saying that this is not a reasonable change. Quite the opposite, it looks reasonable and needed. My point is that we need to look at encoding support holistically to make sure we make it right and consistent. We need to consider various use cases and consequences of supporting them in the context of future changes. Please understand that this is a change to the very core and ubiquitous public API that is hard to change later. To give you a concrete example of why this is  important we recently merged a pull request adding encoding property to the Zip task. Now it became apparent that this property should really be called fileNameEncoding. We'll probably deprecate it and eventually remove it which is always disruptive to the users. This is what happens when we make a change without analyzing the bigger picture. 

Creating design document is the part of every new development done by the Gradle team. We'd like to use the same process for pull requests to maintain the level of quality and consistency that our users expect.

**Comment by pioterj** [comment]
For future reference here's a link to the related JIRA: https://issues.gradle.org/browse/GRADLE-1267

**Comment by AlphaHinex** [comment]
I agree with @pioterj to look at the fucking encoding problem holistically, and I'm so happy that our contributions make Gradle plan to face this now. But as everyone knows (maybe), google is not exist in my country...

**Comment by pioterj** [comment]
It's something that we'd definitely like to improve in Gradle but I don't know when we'll have the capacity to design it ourselves. If any of you would like to try to do it yourself I will make myself available to support it.

**Comment by rjernst** [comment]
@pioterj Thank you for taking the time to look at this issue. I understand the need and desire to develop APIs holistically. However, in this case I think the "design" is very simple: `If a task needs to convert from bytes to characters, or characters to bytes, then it needs an encoding parameter`. This PR does exactly that for the copy task. It is a shame it has sat for so long, while the existing behavior is clearly broken.

**Comment by pioterj** [comment]
@rjernst This is likely to be a sensible piece of a larger pie but I think we need to take a wider look at encoding support in Gradle first. This will allow us to validate that this change is in line with a more holistic solution. We do a lot of research, spiking and deep thinking before implementing new things at Gradle. We find it to be a very important part of the process that helps us to maintain high-quality product with a lot of public APIs. This may look like a simple change but it's in the very core of Gradle and consequences of making it this particular way are not immediately obvious.

**Comment by eljobe** [comment]
@jnizet @rjernst @AlphaHinex I am really interested in helping you get support for content encodings other than the platform default integrated into Gradle's codebase. In fact, this PR looks like a pretty great step in the right direction. However, @pioterj is totally correct to ask for a design spec which explains generally how this change fits into a wholistic approach to getting content encoding correct for all Gradle tasks.

I see @rjernst has taken a stab at the beginning of the design spec here in the comment thread stating:
`If a task needs to convert from bytes to characters, or characters to bytes, then it needs an encoding parameter` However, this is not really a design spec as much as a really good problem statement. For instance it doesn't explain how we keep the name of the encoding parameter clear and consistent across any tasks which convert from bytes to characters. It doesn't enumerate the current set of tasks which fall into this category and a suggested plan for testing the functionality as it is added to each task.

Please, pretty-please, just put in the effort of adding a [design doc](https://github.com/gradle/gradle/tree/master/design-docs) to the repository in a separate pull request so we can get this PR merged and the bug closed. You're so close, don't give up now!

If no one can commit to laying out the design by the end of this week, we're going to have to close this PR without merging it. We just don't have the resources in the core team to write the design right now.  Please give it a shot. I promise it won't hurt as much as you think.

**Comment by AlphaHinex** [comment]
Is there possible to use the same property for the file name and content encoding? I think most of time they are the same, and it's easy to change file name encoding same as content's.

**Comment by eljobe** [comment]
That's an interesting possibility. Why not suggest it in the design doc, and we'll talk it over in that review? It will be a better venue for discussing these details where everyone interested in the future of our public API can weigh in and make sure we come up with the best solution.

**Comment by eskatos** [comment]
> Is there possible to use the same property for the file name and content encoding? I think most of time they are the same, and it's easy to change file name encoding same as content's.

I'm not sure they are the same most of the time.

If you are into setting encoding by operation and not globally I guess this is because you are dealing with several encodings, working with files that are using an encoding that is different from the global one. In such cases I guess that the filename encoding on the filesystem will be the global one most of the time, not the same as the content.

**Comment by eljobe** [comment]
Well, it is a real shame that nobody has been able to take up the design work required to move forward with this pull request. It's been two weeks with no movement on this Pull Request from the community, so I think we have to assume the commitment to getting this change into the Gradle code base is just not there.  I'll close this Pull Request and the related one #506 about the Tar task tomorrow if I we cannot find someone to champion this effort for the community.

:disappointed:

**Comment by rjernst** [comment]
@eljobe Closing a perfectly good PR does not make any sense. Once again, the rule is utterly simple:

```If a task needs to convert from bytes to characters, or characters to bytes, then it needs an encoding parameter```

Look at ANY java method (or even groovy) that deals with bytes/characters. They all have optional args to take the encoding. This is file-handling-101. To close this PR because someone would not waste time on an archaic contribution process is ridiculous.

**Comment by eljobe** [comment]
@rjernst I very much understand your frustration. One thing I want to say up front. Nobody I have talked to about this Pull Request has any reservations about the fact that a mechanism for specifying encoding is necessary. Most would even say it is overdue. The problem is well-established and well-understood. We need to focus our energy on *how* to solve the problem, and not further establishing that there *is* a problem.

I'm really sad that our request for a design document is viewed by even one contributor as part of "an archaic contribution process." In fact, many contributions are accepted without an accompanying design document. We try to make the process as light-weight as possible, and only ask for design work when we feel like it is absolutely called for.

Perhaps the problem is that we haven't clearly articulated why we are asking for a design document on this particular change. Basically, it boils down to this. While everyone agrees that the *problem* needs to be solved, there are several open questions about whether this Pull Request, as it currently exists, is the right *solution.*

1. Should the setting be called `encoding` `fileEncoding` `contentEncoding` or something else?
2. How does this setting interact with the JVM's `file.encoding` property?
3. How do we achieve consistency between various tasks? (Zip tasks have an `encoding` property which is used for encoding/decoding file names.)
4. Should we have separate settings for file name encoding and content encoding?
5. What is the test and documentation plan for the new feature?

Please fight any urge you might have to answer those questions right here in the "Conversation" tab of the Pull Request. If we answer them here, we will lose the record of the discussion for future engineers who are trying to reason about how Gradle handles encoding. That's what design docs are for. Design documents are meant to be a tool for getting groups of engineers to a shared understanding, and capturing decisions about the evolution of the codebase.

You might be thinking, "But why close the Pull Request?" This is a slightly complex answer. Let me give some context. When people come to a project like ours and want to contribute a change, they immediately start to evaluate how likely their change is to be welcomed and eventually merged into the code-base. One of the many signals is how many open Pull Requests are in the queue. The fewer the better. Of course, another signal is how many Pull Requests are Closed having not been merged into the codebase. Again, the fewer the better. So, we are really motivated to accept and merge Pull Requests, and we are motivated to do it in a timely manner. But, that involves holding both ourselves and our contributors responsible for continuing to make progress on work at hand. In the case of this Pull Request, the original contributor (hi, @jnizet) hasn't responded to the multiple requests for a design document since December 2nd of last year. As maintainers of this project, we are happy to work with enthusiastic and responsive contributors to help get their changes into the codebase. But, when contributors stop responding, we are left with one of three choices.

1. Give up on the unfinished Pull Request and Close it without merging.
2. Take on the additional work ourselves in order to resolve the outstanding problems with the request.
3. Just merge the incomplete Pull Request

There are too few of us for option 2 to be a good strategy. We have to be able to rely on healthy collaboration with the community to drive improvements forward. Option 3 is not acceptable because we need to keep the standards of acceptable contributions high or our code quality will suffer and our project will be harder for future contributors to add to. We are left with no choice but to close the Pull Request to keep our house tidy. I really want to underscore the fact that we do *not* like closing Pull Requests without merging them. It hurts the reputation of our project, and it usually (as in this case) means that some valuable enhancement to the project has been abandoned. However, we also want to keep our code quality high, and our project healthy.

If you are motivated to see this change or one like it added to Gradle, we are truthfully committed to guiding your efforts to completion for the good of all Gradle users. Please write a design doc capturing the answers to the questions above in another Pull Request, and we'll get this ball rolling again.

**Comment by jnizet** [comment]
As the original author of the PR, I am also a bit frustrated about this PR not being merged. And I'm sorry not to have reacted earlier to the request for a document.

The main reasons for that lack of reaction are the following ones
- I'm not sure what is expected in this document, and in which form
- I have the feeling that all the choices regarding naming and defaults are already explained in the initial PR, and are documented
- if the goal is to create a design document covering all the existing tasks and plugins of gradle, I feel completely under-qualified

So, I'm glad you ask concrete questions to which I can answer (or give my opinion):

> Should the setting be called encoding fileEncoding contentEncoding or something else?

My first choice was `encoding`. But then I saw that this name was already used by the Zip task for the encoding of file names in the zip. So I used `fineEncoding`. I agree that `contentEncoding` would be clearer, and I'm ready to refactor the code I submittted if the gradle team finds it to be the best name. The choice is yours.

> How does this setting interact with the JVM's file.encoding property?

It doesn't. If an encoding is specified, it's used. Otherwise, the previous behavior is kept, as indicated in the documentation: *By default, the platform default encoding is used*. 

> How do we achieve consistency between various tasks? (Zip tasks have an encoding property which is used for encoding/decoding file names.)

I'm not aware of all the other tasks where an encoding is already used, and where an encoding **should** be used. So I can't answer that question. I chose **not** to use `encoding` precisely because Zip already has `encoding`. I could have renamed the zip task `encoding` to something like `entryNameEncoding`, but that would have broken backward compatibility, which looked to me like a big no-no. If the gradle team is OK to use a better name than `encoding` and thus break compatibility, I can amend my PR to do that. But I don't think I'm the person qualified to take that decision. The choice is yours.

> Should we have separate settings for file name encoding and content encoding?

I think so, yes. But again, I'll let the gradle team take that decision. And to decide if the file entry encoding should have the platform default, or the specified content encoding by default. The choice is yours.

> What is the test and documentation plan for the new feature?

I'm not sure what you mean by "test plan". The PR has unit tests and integration tests covering the new feature. If they look obscure or not enough, I'm willing to accept critics and amend the PR. But the new code is covered. Same for the documentation.

So, to recap, if what you're expecting is a document (but in which form?) containing the questions you asked, and the answers I just gave, then fine, I can do that. But there are questions I can't and should not answer, because I'm just not the person who has the best overall view on the project, and not the final word on the design either.

**Comment by eljobe** [comment]
@jnizet Welcome back to the discussion. Here are my thoughts.

### On going silent

>    I'm not sure what is expected in this document, and in which form

Rather than allowing your lack of knowledge inhibit your participation, why not ask? Both @pioterj and I linked to our repository of [design docs](https://github.com/gradle/gradle/tree/master/design-docs) earlier in this thread. If it wasn't clear before, we're asking that you make a separate pull request which adds a document like one of those to that directory.

>    I have the feeling that all the choices regarding naming and defaults are already explained in the initial PR, and are documented

Rather than remaining quiet, why not ask why we want a design doc given the level of documentation you've already provided?  The answer is pretty straight-forward. We want to have a canonical place to point others to which will quickly answer their question, "I'm adding a task that converts between bytes and characters, how should it handle encoding?" We want to be able to say, "Over here in this design document you can read our current thinking on content encoding." Instead of, "Oh, didn't somebody make a similar change in the past, let's see if we can't dig up the Pull Request and read all the back-and forth conversation about all the different decisions which were hashed out in that discussion and divine from the documentation in the code which decisions were made and for what reasons."

>    if the goal is to create a design document covering all the existing tasks and plugins of gradle, I feel completely under-qualified

I don't think we need to go that far. We just want to stake a claim for a reasonable solution and lay the blueprint for how tasks should handle encoding. We can identify the places we already know of tasks (e.g. Zip) which run afoul of the blueprint we settle on and mention that there may be others. I think the best thing to do in the design document is layout the problem, and the generic solution you propose including the answer's you've listed below.

My general feedback though is not to let things you don't understand or are frustrated by stop you from engaging in the development work. Just keep asking questions and explaining things until we can get stuff worked out. We aim to be reasonable collaborators, so if you feel like we are being unreasonable, ask us questions.

### How to Proceed

Many of your reasons for not wanting to write a design document is that you claim not to be the right person for the job. You don't have enough context. You don't know the right answers to some of the questions.You think it should be up to the gradle team to make those decisions, not you. I'm paraphrasing, but I think I've captured the sentiment.

We don't expect the author of the design document to have all the answers. The point of writing a design document is to collaboratively come up with the right answers and get them recorded for posterity. When anyone (a maintainer, a contributor) adds a design document to our repository, it is understood that others will look at the document and read the proposed changes and provide feedback and suggestions to make it better. The author of the document just initially suggests one way things could work and commits to iterating on it until we have a good idea of how to proceed.

It seems like there is a lot of resistance to even just starting this process, so I took 45 minutes out of my day and wrote up a [draft of the design doc](https://github.com/gradle/gradle/blob/master/design-docs/encoding.md), I've been begging for the community to write. It is not a finished document. Don't just go off implementing the plan laid out in that design doc. Instead, please take ownership of that document by forking it and making a Pull Request maybe renaming it to `handling-encodings.md` so we get a complete diff to look at and review.

Then, we'll have an appropriate place to have a thorough discussion of how we want to handle encodings in Gradle. Your goal in owning that document will just be to update the document to capture the decisions made by the team while discussing it.

If you're up for it, we can reopen this Pull Request and the one for the Tar task once the one for the design work is under way.

And as always, __please__ let me know if you have any questions or concerns with this process. Don't just stop working on things.

**Comment by jnizet** [comment]
@eljobe @pioterj @rjernst I created a design doc, starting from the draft, in #599.

**Comment by eljobe** [comment]
Thanks for the document. I'm have some other folks from the team take a look a the review, but I'll go ahead and reopen this PR so we will have it ready when the review is done.

**Comment by jnizet** [comment]
I rebased on master, changed the name of the property to `filteringCharset` as specified in the design docs, made the build throw an exception if the `filteringCharset` is set to null or to an invalid charset as specified in the design docs, and improved the tests accordingly. 

This PR is now, I think, conform to the design docs, and can thus be reviewed.

File: subprojects/core/src/main/groovy/org/gradle/api/file/CopySpec.java [hunk 1]
```diff
@@ -324,4 +324,17 @@
...
+     * By default, the platform default encoding is used.
+     */
+    String getFilteringCharset();
+
+    /**
+     * Specifies the character encoding used when filtering files, to read and rewrite them.
+     *
+     * @param charset the name of the charset to use when filtering files
+     */
+    void setFilteringCharset(String charset);
```
Comment (eskatos) [inline subprojects/core/src/main/groovy/org/gradle/api/file/CopySpec.java]: Let's use *charset* everywhere instead of mixing *charset* and *encoding*.
```
Gets the charset used to read and write files when filtering.
By default, the JVM default charset is used.
```
and
```
Specifies the charset used to read and write files when filtering.
```

File: subprojects/core/src/main/groovy/org/gradle/api/internal/file/copy/FilterChain.java [hunk 2]
```diff
@@ -23,14 +23,25 @@
...
     private final ChainingTransformer<Reader> transformers = new ChainingTransformer<Reader>(Reader.class);
+    private final String fileEncoding;
+
+    public FilterChain() {
+        this(null);
+    }
+
+    public FilterChain(String encoding) {
+        this.fileEncoding = encoding == null ? Charset.defaultCharset().name() : encoding;
+    }
```
Comment (eskatos) [inline subprojects/core/src/main/groovy/org/gradle/api/internal/file/copy/FilterChain.java]: Same here, is should be `charset`, not `encoding` nor `fileEncoding`.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/file/copy/FilterChainTest.java [hunk 3]
```diff
@@ -1,89 +1,109 @@
...
+    public void canFilterUsingISO88591() throws IOException {
+        canFilterUsingEncoding("ISO_8859_1");
+    }
+
+    @Test
+    public void canFilterUsingUTF8() throws IOException {
+        canFilterUsingEncoding("UTF8");
+    }
+
+    private void canFilterUsingEncoding(String encoding) throws IOException {
```
Comment (eskatos) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/file/copy/FilterChainTest.java]: `s/encoding/charset/i` in the method, parameter and variable names

**Comment by eskatos** [comment]
@jnizet I reviewed the changes. They looks good to me. I added some comments about minor naming issues, could you please fix theses?

One remaining thing would be to add some words in the user guide.
I'm thinking about the *Filtering files* section in the *Working with files* chapter.
See https://docs.gradle.org/2.12/userguide/working_with_files.html#N115CC

The content sits in `subprojects/docs/src/docs/userguide/workingWithFiles.xml`.
The included `filter-files` snippet can be found in `subprojects/docs/src/samples/userguide/files/copy/build.gradle`.
You can build the user guide and preview your changes using:
```
./gradlew docs:userguideHtml
open subprojects/docs/build/docs/userguide/working_with_files.html
```
Let me know if you need anything.

**Comment by jnizet** [comment]
Of course. I'm sorry about the bad naming. I thought I had made all the changes, but I must have messed up somewhere. I'll do the fixes and add some words in the user guide.

**Comment by eskatos** [comment]
No problem. Thanks!

**Comment by jnizet** [comment]
@eskatos done.

**Comment by LeifRilbe** [comment]
I would just like to add that since there is both reading and eriting going on in the copy task, there really should be two encoding parameters. One for reading and one for writing.
